File: java/src/main/java/com/experian/dvs/client/address/layout/elements/ElementLibrary.java
package com.experian.dvs.client.address.layout.elements;

import com.experian.dvs.client.address.Dataset;
import com.experian.dvs.client.address.layout.AppliesTo;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.EnumMap;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public abstract class ElementLibrary {

    private ElementLibrary() {
        throw new IllegalStateException("Utility class");
    }

    private static final Map<Dataset, Class<? extends AddressElement>> DATASET_TO_ADDRESS_ELEMENT_CLASS_MAP = Map.of(
            Dataset.AU_ADDRESS, Aus.class,
            Dataset.AU_ADDRESS_GNAF, Aug.class
    );

    private static final Map<Dataset, Map<String, AddressElement>> ELEMENT_NAME_TO_ADDRESS_ELEMENT_MAP = new EnumMap<>(Dataset.class);

    private static Map<String, AddressElement> getElementNameToAddressElementMap(final Dataset dataset) {

        var cls = DATASET_TO_ADDRESS_ELEMENT_CLASS_MAP.get(dataset);
        if (cls == null) {
            throw new IllegalArgumentException("No AddressElements class found for dataset: " + dataset.name());
        }
        try {
            final Method valuesMethod = cls.getMethod("values");
            final Object[] enumConstants = (Object[]) valuesMethod.invoke(null);
            final Map<String, AddressElement> elementNameToAddressElementMap = new HashMap<>();
            for (Object constant : enumConstants) {
                final Method getElementNameMethod = constant.getClass().getMethod("getElementName");
                final String elementName = (String) getElementNameMethod.invoke(constant);
                elementNameToAddressElementMap.put(elementName, (AddressElement) constant);
            }
            return elementNameToAddressElementMap;
        } catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
            throw new IllegalStateException("Error looking up address elements for dataset " + dataset.name(), e);
        }
    }

    public static AddressElement getAddressElementFromElementName(List<AppliesTo> appliesTo, final String elementName) {
        for (AppliesTo nextAppliesTo : appliesTo) {
            for (Dataset dataset : nextAppliesTo.getDatasets()) {
                AddressElement addressElement = getAddressElementFromElementName(dataset, elementName);
                if (addressElement != null) {
                    return addressElement;
                }
            }
        }
        return null;
    }


    public static AddressElement getAddressElementFromElementName(final Dataset dataset, final String elementName) {
        final Map<String, AddressElement> elementNameToAddressElementMap = ELEMENT_NAME_TO_ADDRESS_ELEMENT_MAP.computeIfAbsent(dataset, ElementLibrary::getElementNameToAddressElementMap);
        return elementNameToAddressElementMap.get(elementName);
    }
}

File: csharp/DVSClient/DVSClient/address/layout/elements/ElementLibrary.cs
using System.Reflection;

namespace DVSClient.Address.Layout.Elements
{
    public abstract class ElementLibrary
    {
        private ElementLibrary()
        {
            throw new InvalidOperationException("Utility class");
        }

        private static readonly Dictionary<Dataset, Type> DatasetToAddressElementClassMap = new Dictionary<Dataset, Type>
        {
            { Dataset.AuAddress, typeof(Aus) },
            { Dataset.AuAddressGnaf, typeof(Aug) },
            { Dataset.GbAddress, typeof(Gbr) },
        };

        private static readonly Dictionary<Dataset, Dictionary<string, IAddressElement>> ElementNameToAddressElementMap = new Dictionary<Dataset, Dictionary<string, IAddressElement>>();

        private static Dictionary<string, IAddressElement> GetElementNameToAddressElementMap(Dataset dataset)
        {
            if (!DatasetToAddressElementClassMap.TryGetValue(dataset, out var cls))
            {
                throw new ArgumentException($"No AddressElements class found for dataset: {dataset.DatasetCode}");
            }

            try
            {
                if (cls != null)
                {
                    var fields = cls.GetFields(BindingFlags.Static | BindingFlags.Public);
                    var elementNameToAddressElementMap = new Dictionary<string, IAddressElement>();

                    MethodInfo? methodInfo = cls.GetMethod("GetElement");

                    if (fields != null && methodInfo != null)
                    {
                        for (int i = 0; i < fields.Length; i++)
                        {
                            IAddressElement? info = (IAddressElement?)methodInfo.Invoke(methodInfo, new object[] { fields[i].Name });
                            if (info != null)
                            {
                                elementNameToAddressElementMap[info.GetElementName()] = info;
                            }
                        }

                        return elementNameToAddressElementMap;
                    }
                }

                throw new InvalidOperationException($"Error looking up address elements for dataset {dataset}");
            }
            catch (Exception e) when (e is TargetInvocationException || e is MethodAccessException)
            {
                throw new InvalidOperationException($"Error looking up address elements for dataset {dataset}", e);
            }
        }

        public static IAddressElement? GetAddressElementFromElementName(IEnumerable<AppliesTo> appliesTo, string? elementName)
        {
            foreach (var nextAppliesTo in appliesTo)
            {
                foreach (var dataset in nextAppliesTo.Datasets)
                {
                    var addressElement = GetAddressElementFromElementName(dataset, elementName);
                    if (addressElement != null)
                    {
                        return addressElement;
                    }
                }
            }
            return null;
        }

        public static IAddressElement? GetAddressElementFromElementName(Dataset dataset, string? elementName)
        {
            ElementNameToAddressElementMap.TryAdd(dataset, GetElementNameToAddressElementMap(dataset));
            ElementNameToAddressElementMap.TryGetValue(dataset, out var dictionaryEntry);

            if (dictionaryEntry != null && !string.IsNullOrWhiteSpace(elementName))
            {
                return dictionaryEntry.TryGetValue(elementName, out var addressElement) ? addressElement : null;
            }

            return null;
        }
    }
}